`Refinery`: Transformation for escaping html/css/js

lscharmer opened the pull request:
This PR introduces 5 new transformations to escape: Javascript, HTML, HTML attributes, CSS and URL's.
The implementation is based upon this repository: https://github.com/laminas/laminas-escaper/

mjansenDatabay commented on src/Refinery/Escape/Transformation/HTMLAttribute.php:
```diff
@@ -0,0 +1,72 @@
+    private function replace(string $chr): string
+    {
+        if ($this->isNonPrintableControl($chr)) {
+            return '&#xFFFD;';
```
Could you add a comment which describes what `&#xFFFD;` is?

mjansenDatabay commented on src/Refinery/Escape/Transformation/HTMLAttribute.php:
```diff
@@ -0,0 +1,72 @@
+        };
+    }
+
+    public function isNonPrintableControl($chr): bool
```
Why not `private`?

lscharmer replied:
Could question :)

klees reviewed (changes requested):
Hi @lscharmer,

please be so kind and try to provide minimal PRs. In this case, one PR per Transformation would be nice. This would improve various things:

* I could look into PRs incrementally or even ask other people to look into smaller PRs.
* We could merge PRs incrementally. I guess I could have merged some of the changes already without further ado, but now they are all tied together.
* Smaller PRs will be easier to review because I reviewers aren't repelled by the amount of code to be reviewed.

These, and maybe even more, are the reasons we specifically ask to make small PRs in our [rules for contributors](https://github.com/ILIAS-eLearning/ILIAS/blob/release_8/docs/development/contributing.md#rules-for-contributors). There are times when big PRs are warranted. This is not one of these times.

So, now to the content =)

Please answer the following questions:
- [x] CSS: Why do we need this transformation? Do we have a usecase for it?
- [x] HTMLAttributes: Why do we need this transformation? Do we have a usecase for it?
- [x] JavaScript: Why do we need this transformation? Do we have a usecase for it?
- [x] u: What does the `u` modifier in your regexps do? [This](https://github.com/ILIAS-eLearning/ILIAS/pull/6314/files#diff-92e0d450d8ec63ba77a6e87023294d0d2a6d9f061728da368f85befca262de78R39) seems to suggest that it somehow only matches UTF-8 strings, but I could not find according docu. Could you link some? Could you give me an example of a string that would make that expression fail?

Please implement the following changes:
- [x] Wording Group: I think what we do here is "Encoding" not "Escaping". "Escaping" means to have a certain character (say `\`) that makes the next character(s) behave differently than normal (make `\n` a linebreak). We "escape the n". Encoding on the other hand is about how data is represented. This is what we are doing here: Move data in UFT8-space to special character sequences that might contain escape characters. I would propose to rename the group to "Encode".
- [x] Wording HTML: Please rename the transformation to "HTMLSpecialCharsAsEntities" 
- [x] Wording CSS: Please rename the transformation to "NonAlphaNumToCSSEscapes" 
- [x] Docu CSS: Please document what [this transformation](https://github.com/ILIAS-eLearning/ILIAS/pull/6314/files#diff-a6b677866a51479f2ca08ef1f729a9de1a7fc21b25cc9018f5ef27df19d1a72fR32) is supposed to do. Maybe link some source such as [this](https://www.w3.org/International/questions/qa-escapes#cssescapes).
- [x] Docu HTML: Please document what [this transformation](https://github.com/ILIAS-eLearning/ILIAS/pull/6314/files#diff-92e0d450d8ec63ba77a6e87023294d0d2a6d9f061728da368f85befca262de78R32) is supposed to do. Maybe link some sources.
- [x] Docu URL: Same as for HTML.
- [x] Docu HTMLAttributes: Please document what [this transformation](https://github.com/ILIAS-eLearning/ILIAS/pull/6314/files#diff-b83583f7eb3a52d71a791c02009ccfe2a3a47bbeb67bd969b324d3105dc2997dR30) is supposed to do. Maybe link some sources. Judging on the examples in the tests, this transformation seem to do various things at once and, frankly, I cannot grasp how and where I would want to use it or what it is even supposed to do. I might ask to iterate on this further.
- [ ] Docu JavaScript: Same as for `HTMLAttributes` 
- [ ] Docu UTF8Chars: Please document what the methods in that trait are supposed to do.

In general, there are many things here that I currently do not understand here. Also, it took me quite some time to look through this. Besides smaller PRs a better exposition would have helped a lot: Explain stuff in the the summary of the PR, add docu to the classes and maybe methods you expose, give names that are suitable or even can be used as a starting point to search for stuff on the internet.

Kind regards!

klees commented:
Hi @lscharmer,

I will assign this to you, so this does not appear on my agenda anymore.

Please assign back if this is ready for the next iteration.

Kind regards!

lscharmer commented:
Hi @klees,

I updated this PR and addressed your points:

>     * [ ]  CSS: Why do we need this transformation? Do we have a usecase for it?
In agreement with @mjansen, we decided that we don't and it is therefore removed from this PR :)

>     * [ ]  HTMLAttributes: Why do we need this transformation? Do we have a usecase for it?
There are many places where PHP variables are passed to templates where they are used as HTML attributes.
E.g. https://github.com/ILIAS-eLearning/ILIAS/blob/0514bac85249b86002d05915fcab7427e15d3a63/components/ILIAS/UI/src/Implementation/Component/Link/Renderer.php#L74 would be one use case where the value should be escaped as a HTML attribute.

>     * [ ]  JavaScript: Why do we need this transformation? Do we have a usecase for it?
This is useful in places where JavaScriptBindable::withAdditionalOnLoadCode(...) is used to add JS from PHP.
E.g. Here: https://github.com/ILIAS-eLearning/ILIAS/blob/0514bac85249b86002d05915fcab7427e15d3a63/components/ILIAS/UI/src/Implementation/Component/Button/Renderer.php#L94

>     * [ ]  u: What does the `u` modifier in your regexps do? [This](https://github.com/ILIAS-eLearning/ILIAS/pull/6314/files#diff-92e0d450d8ec63ba77a6e87023294d0d2a6d9f061728da368f85befca262de78R39) seems to suggest that it somehow only matches UTF-8 strings, but I could not find according docu. Could you link some? Could you give me an example of a string that would make that expression fail?
The `u` modifier is very important. The behaviour of the modifier is documented here: https://www.php.net/manual/en/reference.pcre.pattern.modifiers.php
Additionally the first user contributed note on that page describes the behaviour with invalid UTF-8 strings with this modifier in detail.
I use the `u` modifier to ensure that the given string is a valid UTF-8 string. IMO this is the most fail prove way to check for a valid UTF-8 string, as it is handled by PHP itself, instead of implementing this oneself. If you know a better way to check for a valid UTF-8 string I will change this of course.
The following expression will fail: `preg_match('//u', "\x80")` but without the `u` modifier it will not.

>     * [ ]  Wording Group: I think what we do here is "Encoding" not "Escaping". "Escaping" means to have a certain character (say `\`) that makes the next character(s) behave differently than normal (make `\n` a linebreak). We "escape the n". Encoding on the other hand is about how data is represented. This is what we are doing here: Move data in UFT8-space to special character sequences that might contain escape characters. I would propose to rename the group to "Encode".
Done.

>     * [ ]  Wording HTML: Please rename the transformation to "HTMLSpecialCharsAsEntities"
Done. I hope the documentation and links I added are enough, as this is just a wrapper around a builtin PHP function.

>     * [ ]  Wording CSS: Please rename the transformation to "NonAlphaNumToCSSEscapes"
Is removed instead.

>     * [ ]  Docu CSS: Please document what [this transformation](https://github.com/ILIAS-eLearning/ILIAS/pull/6314/files#diff-a6b677866a51479f2ca08ef1f729a9de1a7fc21b25cc9018f5ef27df19d1a72fR32) is supposed to do. Maybe link some source such as [this](https://www.w3.org/International/questions/qa-escapes#cssescapes).
Is removed instead.

>     * [ ]  Docu HTML: Please document what [this transformation](https://github.com/ILIAS-eLearning/ILIAS/pull/6314/files#diff-92e0d450d8ec63ba77a6e87023294d0d2a6d9f061728da368f85befca262de78R32) is supposed to do. Maybe link some sources.
Done. I hope the documentation and links I added are enough, as this is just a wrapper around a builtin PHP function.

>     * [ ]  Docu URL: Same as for HTML.
Done. I hope the documentation and links I added are enough, as this is just a wrapper around a builtin PHP function.

>     * [ ]  Docu HTMLAttributes: Please document what [this transformation](https://github.com/ILIAS-eLearning/ILIAS/pull/6314/files#diff-b83583f7eb3a52d71a791c02009ccfe2a3a47bbeb67bd969b324d3105dc2997dR30) is supposed to do. Maybe link some sources. Judging on the examples in the tests, this transformation seem to do various things at once and, frankly, I cannot grasp how and where I would want to use it or what it is even supposed to do. I might ask to iterate on this further.
Done, this class is renamed to `HTMLAttributeValue`. This transformation only encodes a given string to be a valid HTML attribute value. This just includes all of the requirements of the HTML spec.
There are just a lot of requirements.
I add documentation to elaborate on all of the requirements of an HTML attribute value and added links to the corresponding HTML specs where necessary and I renamed all variables to be more fitting to the domain, so one can search for these words online.

>     * [ ]  Docu JavaScript: Same as for `HTMLAttributes`
After some research I noticed that this can be implemented much simpler. It uses `json_encode` with correct flags now.
The transformation is renamed to `JsonEncode`.

>     * [ ]  Docu UTF8Chars: Please document what the methods in that trait are supposed to do.
Done, this trait is also renamed to `ProvideUTF8CodepointRange`.



Best regards
@lscharmer

alex40724 commented:
@lscharmer Thanks for this PR. This is definitely needed. I would strongly suggest to extend the readme file https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/components/ILIAS/Refinery/README.md with this PR, too, since this is what most devs will read. Outlining the main use cases would be very helpful.

lscharmer committed:
Refinery: Add transformation for escaping html/css/js/url's

lscharmer commented:
> @lscharmer Thanks for this PR. This is definitely needed. I would strongly suggest to extend the readme file https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/components/ILIAS/Refinery/README.md with this PR, too, since this is what most devs will read. Outlining the main use cases would be very helpful.

Hi @alex40724,
the `README.md` is adjusted accordingly.

I also renamed the `JsonEncode` transformation to `Json` as `encode` is already in the group / path name (`$refinery->enode()->json()`).

mjansenDatabay commented:
I added this to the JF agenda for educational purposes :-).

alex40724 commented:
@lscharmer Thanks!

matthiaskunkel commented:
Jour Fixe, 27 MAY 2024: @mjansenDatabay notified us about the transformations for escaping html/css/js and asks all developers to use this methods for ILIAS 10. PR will be merged to trunk.

nicoroeser commented on components/ILIAS/Refinery/src/Encode/Transformation/HTMLAttributeValue.php:
```diff
@@ -0,0 +1,107 @@
+    {
+        return strlen($utf8_char) === 1 ?
+            ctype_cntrl($utf8_char) && !ctype_space($utf8_char) :
+            0x1F <= $codepoint && $codepoint <= 0x9F;
```
Nitpick: the `0x1F <= $codepoint && ` part is a bit confusing after you have read the function comment. It does not matter, but it will always be true; we could remove it. Or replace `0x1F` with `0x7F` (which would make the check less strict, but still does not matter).

If I’m mistaken, please enlighten me; if not, consider dropping the part from the code. Thanks. :smiley:

lscharmer replied:
You are correct, I will remove that part. Thanks for the correction :)

lscharmer replied:
Is removed

nicoroeser replied:
Thanks! :smiley:
